### PR TITLE
use taipy resource handler route to resolve cookies

### PR DESCRIPTION
**Pull request by dinhlongviolin1**

### CODE REVIEW

**Review by FredLL-Avaiga — DISMISSED** [review]
Good start @dinhlongviolin1

**Review by FabienLelaquais — DISMISSED** [review]
🤐

File: frontend/taipy-gui/base/src/cookieHandler.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    async init(socket: Socket, taipyApp: TaipyApp) {
+        const hasValidCookies = await this.verifyCookieStatus();
+        if (!hasValidCookies) {
+            await this.deleteCookie();
+            localStorage.removeItem(TAIPY_RESOURCE_HANDLER);
+            window.location.reload();
+            return;
+        }
+        this.addBeforeUnloadListener();
+        initSocket(socket, taipyApp);
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/cookieHandler.ts:23]: Not sure about the init Socket inside the CookieHandler ...

File: frontend/taipy-gui/base/src/cookieHandler.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import { TaipyApp } from "./app";
+import { initSocket } from "./socket";
+import axios from "axios";
+import { getLocalStorageValue } from "../../src/context/utils";
+
+export const TAIPY_RESOURCE_HANDLER = "tprh";
+
+declare global {
+    interface Window {
+        deleteCookie: () => void;
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/cookieHandler.ts]: is this declaration not available by default ?

File: frontend/taipy-gui/base/src/cookieHandler.ts (hunk 1 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/base/src/cookieHandler.ts:23]: especially if hasValidCookies is false
Then initSocket is not called at all !

File: frontend/taipy-gui/base/src/cookieHandler.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        // If not, then some part of the code must have removed the cookie
+        // or wants to remove the cookie by removing the local storage
+        if (!this.resourceHandlerId) {
+            return new Promise((resolve) => resolve(false));
+        }
+        try {
+            // call to get cookie status
+            const { data } = await axios.get("taipy-rh");
+            // validate cookie status
+            if (!data || !data.rh_id || data.rh_id !== this.resourceHandlerId) {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/cookieHandler.ts]: could this be if (data?.rh_id !== this.resourceHandlerId)  ?

File: frontend/taipy-gui/base/src/cookieHandler.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            }
+        } catch (error) {
+            console.error("Error while validating cookie:", error);
+            return new Promise((resolve) => resolve(false));
+        }
+        return new Promise((resolve) => resolve(true));
+    }
+
+    addBeforeUnloadListener() {
+        window.deleteCookie = () => {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/cookieHandler.ts]: why do you need to make it an window attribute ?

File: frontend/taipy-gui/base/src/cookieHandler.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            axios.delete("taipy-rh");
+        };
+        window.addEventListener("beforeunload", () => {
+            localStorage.removeItem(TAIPY_RESOURCE_HANDLER);
+            window.deleteCookie();
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/cookieHandler.ts]: why not use `self.deleteCookie()`
with `const self = this;` before the addEventListener

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
tehre's a case when initSocket will not be called

File: frontend/taipy-gui/base/src/cookieHandler.ts (hunk 2 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/cookieHandler.ts]: we have to make this declaration as I will be using it during the before unload

File: frontend/taipy-gui/base/src/cookieHandler.ts (hunk 1 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/cookieHandler.ts:23]: We dont need to call init socket in this case as the cookie is invalid. The key is that we will get out of this instance and go back to taipy-gui to get a new cookie. I put it here and do things asynchronously as I dont want to start the websocket connection until i am sure the cookie is valid.

File: frontend/taipy-gui/base/src/cookieHandler.ts (hunk 3 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/cookieHandler.ts]: make sense to me thanks!

File: frontend/taipy-gui/base/src/cookieHandler.ts (hunk 5 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/cookieHandler.ts]: this.deleteCookie().bind(this) should do the trick but I just want to make sure that the window does not have any dependency on the teardown function

**Review by FabienLelaquais — APPROVED** [review]
👍

File: frontend/taipy-gui/base/src/cookieHandler.ts [hunk 6]
```diff
@@ -51,12 +45,9 @@ export class CookieHandler {
     }
 
     addBeforeUnloadListener() {
-        window.deleteCookie = () => {
-            axios.delete("taipy-rh");
-        };
         window.addEventListener("beforeunload", () => {
             localStorage.removeItem(TAIPY_RESOURCE_HANDLER);
-            window.deleteCookie();
+            this.deleteCookie();
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/cookieHandler.ts:50]: are you sure `this` is valid in this context ?
I would have saved it in the scope with `const self = this;`

**Review by FredLL-Avaiga — APPROVED** [review]

File: frontend/taipy-gui/base/src/cookieHandler.ts (hunk 6 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/base/src/cookieHandler.ts:50]: I think arrow function (method) are automatically bound ?

Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/cookieHandler.ts:50]: It is still working per my testing. I will try your method later on if it breaks somehow and I like your proposal very much too!

**Pull request merged**